lexers/meson.lua

```diff
@@ -1,23 +1,60 @@
 -- Copyright 2020-2023 Florian Fischer. See LICENSE.
 -- Meson file LPeg lexer.
 
-local lexer = require('lexer')
-local token, word_match = lexer.token, lexer.word_match
-local P, R, S = lpeg.P, lpeg.R, lpeg.S
+local lexer = lexer
+local S = lpeg.S
 
-local lex = lexer.new('meson', {fold_by_indentation = true})
-
--- Whitespace.
-lex:add_rule('whitespace', token(lexer.WHITESPACE, lexer.space^1))
+local lex = lexer.new(..., {fold_by_indentation = true})
 
 -- Keywords.
-lex:add_rule('keyword', token(lexer.KEYWORD, word_match(
-  'and or not if elif else endif foreach break continue endforeach')))
+lex:add_rule('keyword', lex:tag(lexer.KEYWORD, lex:word_match(lexer.KEYWORD)))
 
 -- Methods.
 -- https://mesonbuild.com/Reference-manual.html#builtin-objects
 -- https://mesonbuild.com/Reference-manual.html#returned-objects
-local method_names = word_match{
+-- A method call must be followed by an opening parenthesis.
+lex:add_rule('method', lex:tag(lexer.FUNCTION_METHOD, lex:word_match(lexer.FUNCTION_METHOD) * #(lexer.space^0 * '(')))
+
+-- Function.
+-- https://mesonbuild.com/Reference-manual.html#functions
+-- A function call must be followed by an opening parenthesis. The matching of function calls
+-- instead of just their names is needed to not falsely highlight function names which can also
+-- be keyword arguments. For example 'include_directories' can be a function call itself or a
+-- keyword argument of an 'executable' or 'library' function call.
+lex:add_rule('function', lex:tag(lexer.FUNCTION, lex:word_match(lexer.FUNCTION) * #(lexer.space^0 * '(')))
+
+-- Builtin objects.
+-- https://mesonbuild.com/Reference-manual.html#builtin-objects
+lex:add_rule('object', lex:tag(lexer.VARIABLE_BUILTIN, lex:word_match(lexer.VARIABLE_BUILTIN)))
+
+-- Constants.
+lex:add_rule('constant', lex:tag(lexer.CONSTANT, lex:word_match(lexer.CONSTANT)))
+
+-- Identifiers.
+lex:add_rule('identifier', lex:tag(lexer.IDENTIFIER, lexer.word))
+
+-- Strings.
+local str = lexer.range("'", true)
+local multiline_str = lexer.range("'''")
+lex:add_rule('string', lex:tag(lexer.STRING, multiline_str + str))
+
+-- Comments.
+lex:add_rule('comment', lex:tag(lexer.COMMENT, lexer.to_eol('#', true)))
+
+-- Numbers.
+local oct = '0o' * lexer.oct_num
```

Comment:
This will not match `0o7`. `lexer.oct_num` needs a `0` prefix, so this should be `0o * lpeg.R('07')`.